File: src/main/java/main/utils/resume/ResumableTrack.java
package main.utils.resume;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.topisenpai.lavasrc.mirror.MirroringAudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import main.audiohandlers.TrackScheduler;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.json.JSONArray;
import org.json.JSONObject;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@Slf4j
public class ResumableTrack {
    private final static ObjectMapper mapper = new ObjectMapper();

    @Getter
    private final AudioTrackInfoWrapper info;
    @Getter
    @Nullable
    private final String artworkURL;
    @Getter
    @Nullable
    private final String isrc;
    @Getter
    @Nullable
    private final TrackScheduler.Requester requester;

    public ResumableTrack() {
        this.artworkURL = null;
        this.isrc = null;
        this.info = null;
        this.requester = null;
    }

    public ResumableTrack(@Nullable String artworkURL, @Nullable String isrc, AudioTrackInfoWrapper info, @Nullable TrackScheduler.Requester requester) {
        this.artworkURL = artworkURL;
        this.isrc = isrc;
        this.info = info;
        this.requester = requester;
    }

    public ResumableTrack(AudioTrack track, TrackScheduler.Requester requester) {
        this.info = new AudioTrackInfoWrapper(track.getInfo());
        this.requester = requester;
        if (track instanceof MirroringAudioTrack mt) {
            this.artworkURL = mt.getArtworkURL();
            this.isrc = mt.getISRC();
        } else {
            this.artworkURL = null;
            this.isrc = null;
        }
    }

    public ResumableTrack(Pair<AudioTrack, TrackScheduler.Requester> info) {
        this.info = new AudioTrackInfoWrapper(info.getLeft().getInfo());
        this.requester = info.getRight();
        if (info.getLeft() instanceof MirroringAudioTrack mt) {
            this.artworkURL = mt.getArtworkURL();
            this.isrc = mt.getISRC();
        } else {
            this.artworkURL = null;
            this.isrc = null;
        }
    }

    @Override
    public String toString() {
        try {
            return mapper.writeValueAsString(this);
        } catch (JsonProcessingException e) {
            return null;
        }
    }

    public static List<ResumableTrack> audioTracksToResumableTracks(Collection<Pair<AudioTrack, TrackScheduler.Requester>> tracks) {
        return tracks.stream()
                .map(trackPair -> new ResumableTrack(trackPair.getLeft(), trackPair.getRight()))
                .toList();
    }

    public static String collectionToString(Collection<ResumableTrack> tracks) {
        final var arr = new JSONArray();
        tracks.forEach(track -> arr.put(new JSONObject(track.toString())));
        return arr.toString();
    }

    public static String audioTracksToString(Collection<Pair<AudioTrack, TrackScheduler.Requester>> tracks) {
        final var arr = new JSONArray();
        audioTracksToResumableTracks(tracks).forEach(track -> arr.put(new JSONObject(track.toString())));
        return arr.toString();
    }

    public static List<ResumableTrack> stringToList(String array) {
        final var revertedTracks = new ArrayList<ResumableTrack>();

        new JSONArray(array)
                .spliterator()
                .forEachRemaining(obj -> {
                    try {
                        revertedTracks.add(ResumableTrack.fromJSON(obj.toString()));
                    } catch (JsonProcessingException e) {
                        log.error("Error processing JSON", e);
                    }
                });

        return revertedTracks;
    }

    public static ResumableTrack fromJSON(String json) throws JsonProcessingException {
        return mapper.readValue(json, ResumableTrack.class);
    }

    public static class AudioTrackInfoWrapper extends AudioTrackInfo {

        public AudioTrackInfoWrapper() {
            super(null, null, 0, null, false, null);
        }

        public AudioTrackInfoWrapper(AudioTrackInfo info) {
            super(info.title, info.author, info.length, info.identifier, info.isStream, info.uri);
        }
    }
}

File: src/main/kotlin/main/utils/resume/ResumableTrack.kt
//package main.utils.resume
//
//import dev.arbjerg.lavalink.protocol.v4.Track
//import kotlinx.serialization.Serializable
//import kotlinx.serialization.decodeFromString
//import kotlinx.serialization.encodeToString
//import kotlinx.serialization.json.Json
//import main.audiohandlers.models.Requester
//import main.audiohandlers.utils.artworkUrl
//import main.audiohandlers.utils.isrc
//import org.json.JSONArray
//import org.json.JSONObject
//
//@Serializable
//data class ResumableTrack(
//    val info: AudioTrackInfoWrapper,
//    val artworkURL: String?,
//    val isrc: String?,
//    val requester: Requester?
//) {
//    companion object {
//        fun Collection<Pair<Track, Requester>>.toResumableTracks(): List<ResumableTrack> =
//            this.map { trackPair ->
//                val track = trackPair.first
//                val requester = trackPair.second
//                val info = AudioTrackInfoWrapper(track.info)
//                val artworkURL = track.artworkUrl
//                val isrc = track.isrc
//
//                ResumableTrack(info, artworkURL, isrc, requester)
//            }
//
//        fun Collection<ResumableTrack>.string(): String {
//            val arr = JSONArray()
//            this.forEach { track -> arr.put(JSONObject(track.toString())) }
//            return arr.toString()
//        }
//
//        fun fromJSON(json: String): ResumableTrack =
//            Json.decodeFromString(json)
//    }
//
//    constructor(track: Track, requester: Requester?) : this(
//        info = AudioTrackInfoWrapper(track.info),
//        requester = requester,
//        artworkURL = track.artworkUrl,
//        isrc = track.isrc
//    )
//
//    fun toAudioTrack(sourceManager: ResumeSourceManager): Track =
//        ResumeTrack(info.toAudioTrackInfo(), isrc, artworkURL, sourceManager)
//
//    override fun toString(): String =
//        Json.encodeToString(this)
//}
